### DISCUSSION TITLE
🔧 v0.5.1 Hotfix Released - Embedding Autodiff Fix

### DISCUSSION THREAD

**Discussion by kolkov** [born-ml/born#10] [Announcements]
# v0.5.1 Hotfix - Critical Embedding Fix

## What was fixed

`nn.Embedding.Forward()` was using direct Go slice operations instead of backend operations, which meant **embedding weights were completely disconnected from the autodiff tape**. This caused zero gradients for all embedding parameters.

## Impact

**Before v0.5.1**: All models using `nn.Embedding` (transformers, LLMs, etc.) could not train - embeddings never updated.

**After v0.5.1**: Full gradient flow through embedding layers via scatter-add backward pass.

## Upgrade

```bash
go get github.com/born-ml/born@v0.5.1
```

## Technical Details

- Added `Embedding` method to Backend interface
- Implemented in CPU, Mock, and WebGPU backends
- `nn.Embedding.Forward()` now routes through `backend.Embedding()` which records `EmbeddingOp` on tape
- Backward pass uses scatter-add to accumulate gradients for repeated indices

## Full Changelog

https://github.com/born-ml/born/compare/v0.5.0...v0.5.1

---

If you were experiencing training issues with embedding-based models, please upgrade to v0.5.1!
